pkg/apis/core/validation/validation.go

```diff
@@ -3735,12 +3773,26 @@ func ValidatePodUpdate(newPod, oldPod *core.Pod) field.ErrorList {
 		allErrs = append(allErrs, field.Invalid(specPath.Child("activeDeadlineSeconds"), newPod.Spec.ActiveDeadlineSeconds, "must not update from a positive integer to nil value"))
 	}
 
+	if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
+		// reject QOS change attempt
+		oldQOS := qos.GetPodQOS(oldPod)
+		newQOS := qos.GetPodQOS(newPod)
+		if newQOS != oldQOS {
+			allErrs = append(allErrs, field.Invalid(fldPath, newQOS, "Pod QOS is immutable"))
+		}
+	}
+
 	// handle updateable fields by munging those fields prior to deep equal comparison.
 	mungedPod := *newPod
 	// munge spec.containers[*].image
 	var newContainers []core.Container
 	for ix, container := range mungedPod.Spec.Containers {
 		container.Image = oldPod.Spec.Containers[ix].Image
+		if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
+			// allow Resources and ResourcesAllocated fields to be mutable
+			container.Resources = oldPod.Spec.Containers[ix].Resources
+			container.ResourcesAllocated = oldPod.Spec.Containers[ix].ResourcesAllocated
```

Comment:
Does this mean a user can change the resourcesAllocated?

Reply:
user cannot - admission plugin will yell. kubelet can.